### PR TITLE
Allow app component to be wrapped with custom enhancer when rendering

### LINKED ISSUE #4269: feature request: renderPage can pass props to App component directly
- [x] I have searched the [issues](https://github.com/zeit/next.js/issues?q=is%3Aissue) of this repository and believe that this is not a duplicate.

## Current Behavior

currently `material-ui` use `renderPage` to pass `pageContext` to page Component
```js
// _documents.js
const page = ctx.renderPage(Page => props => (
      <Page pageContext={pageContext} {...props} />
  ));
```

so, **every page** components need to handle `pageContext`
```js
// pages/index.js
export default withRoot(Home) // <-- withRoot HOC handle pageContext
```

## Expected Behavior

but i really want '_app.js' handle withRoot or other HOC,
so **every page** components can be more simplier.

currently `_document.js` can use `renderPage` to  pass props to `Page Component` such as  'pageContext' above.
its for style ssr.

but `_document.js` can't send props to `App Component``_app.js`.
u only can modify `App` props by custom it `/page/_app.js`, and then generate App props by getInitialProps function.
https://github.com/zeit/next.js/blob/canary/server/render.js#L79-L88

but, again, `_document.js` can't send props to `_app.js` directly if i want to handle something about style ssr in _app.js, so that i don't need write those same code in every pages

may be Next.js would modify the renderPage function, may be add `appProps` below
```js
const renderPage = (enhancer = Page => Page, appProps) => {
    const app = createElement(App, {
      Component: enhancer(Component),
      router,
      ...props,
      ...appProps
    })
```

### PR DESCRIPTION
**Pull request by albinekb**
Continuation from #4288

Fixes #4269

### CODE REVIEW

File: test/integration/app-document/pages/_document.js [hunk 1]
```diff
@@ -2,8 +2,17 @@ import Document, { Head, Main, NextScript } from 'next/document'
...
   static async getInitialProps (ctx) {
-    const initialProps = await Document.getInitialProps(ctx)
-    return { ...initialProps, customProperty: 'Hello Document' }
+    let options
+
+    const enhanceComponent = Component => (props) => <div><span id='render-page-html'>RENDERED</span><Component {...props} /></div>
+    if (ctx.query.withEnhancer) {
+      options = enhanceComponent
+    } else if (ctx.query.withEnhanceComponent) {
+      options = { enhanceComponent }
```
Comment (timneutkens) [inline test/integration/app-document/pages/_document.js]: This should also test `enhanceApp`

File: server/render.js [hunk 2]
```diff
@@ -81,12 +81,21 @@ async function doRender (req, res, pathname, query, {
...
-  const renderPage = (enhancer = Page => Page) => {
-    const app = createElement(App, {
-      Component: enhancer(Component),
-      router,
-      ...props
-    })
+  const renderPage = (options = Page => Page) => {
+    let EnhancedApp = App
+    let EnhancedComponent = Component
+    if (typeof options === 'function') {
```
Comment (timneutkens) [inline server/render.js]: Lets add a comment here that this is for backwards compatibility

**Review by timneutkens — APPROVED** [review]

**Comment by timneutkens** [comment]
Thanks!

**Comment by Tapppi** [comment]
Thanks so much for picking up the slack while I'm on holiday, this looks great 🙂

**Pull request merged**